### DISCUSSION TITLE
Could not set I2S master clock line with mck keyword

### DISCUSSION THREAD

**Discussion by saraverbeecke** [micropython/micropython#9248] [Libraries & Drivers] [answered]
I'm trying to initialize a I2S object as described in the docs: [https://docs.micropython.org/en/latest/library/machine.I2S.html](https://docs.micropython.org/en/latest/library/machine.I2S.html) but the keyword **mck** seems not to work.
Any ideas?

I'm using an Generic ESP32 board with SGTL5000 Codec:
```
from machine import Pin, I2S
sck_pin = Pin(15)
ws_pin = Pin(16)
sd_pin = Pin(17)
mck_pin = Pin(18)
audio_in = I2S(0, sck=sck_pin, ws=ws_pin, sd=sd_pin, mck=mck_pin, mode=I2S.RX, bits=16, format=I2S.MONO, rate=22050, ibuf=20000)
```

**Comment by jimmo** [comment, marked as answer]
The `mck` argument is only supported on the m.iMX port. We should clarify that in the documentation.

**Comment by karfas** [comment]
Which micropython version do you use ?

The mck keyword is currently not supported by micropython.

Even when #8110 solves an initialization bug with the mck configuration, I can't see anywhere in the ESP32 port that `mck=Pin` gets evaluated as a parameter.  

I'm pretty sure this is a bug.

**Reply by robert-hh** [reply]
As @jimmo said, a mclk signal is not generated by the ESP32 I2S hardware. But you can create a mclk signal using PWM.
See also https://forum.micropython.org/viewtopic.php?f=2&t=12405&p=67381&hilit=I2S+ESP32#p67381

**Reply by karfas** [reply]
No. @jimmo said, "it is not supported".
While I have no idea which hardware variant of the ESP32 can or can't generate mclk, I'm pretty sure it's in the ESP-IDF configurable **for a reason**.
See e.g. https://docs.espressif.com/projects/esp-idf/en/v4.4.2/esp32/api-reference/peripherals/i2s.html

**Reply by robert-hh** [reply]
Ok. So it seems to be not supported by the MP driver. Still you can generate the mclk signal for the SGTL5000 codec by a PWM, channel. Maybe Mike will find the time to update the ESP32 I2S module.

**Reply by saraverbeecke** [reply]
@karfas I'm using the latest build 1.19-1.

Thanks for the answers!

**Reply by miketeachman** [reply]
It's on my todo list to add Master clock support to the other ports, but admittedly it's not high on my agenda, mainly because the most popular I2S devices do not need a Master clock signal.  e.g.  the devices listed [here ](https://github.com/miketeachman/micropython-i2s-examples#i2s-microphone-boards-tested) and [here](https://github.com/miketeachman/micropython-i2s-examples#i2s-dac-and-amplifier-boards-tested)

One challenge for the ESP32 is the varying level of Master clock support for the different ESP-IDF versions that MicroPython supports.  For example, MicroPython supports ESP-IDF versions v4.0.2, v4.1.1, v4.2.2, v4.3.2 and v4.4.  Of these, only v4.4 supports a Master clock output.  So, that would mean adding conditional compilation macros into the I2S code to detect what version of ESP-IDF is being used, to turn on/off the Master clock feature.  These macros in turn drive regression testing to make sure the I2S feature still builds and works correctly when someone uses say ESP-IDF v4.0.2.  As a volunteer contributor with a busy life, this work has somewhat limited appeal, especially when the Master clock feature can be replicated with `machine.PWM`.
